### PR TITLE
Asynchronous snapshot writers

### PR DESCRIPTION
**Pull request by niboshi**
This PR takes over #3401.
We appreciate the great work of @tyohei on this PR. Thank you very much!

Please refer to #3401 for the objective of this PR and previous discussions.

### CODE REVIEW

**Comment by niboshi** [comment]
Despite the comment I left in the previous discussion, I put writer implementations in a new module `chainer.training.extensions.snapshot_writers`.
Comments / suggestions are welcome.

**Comment by niboshi** [comment]
I also integrated new features into existing `extensions.snapshot()` function, instead of introducing a new classs `extensions.Snapshot` (I renamed the class as `_Snapshot`).

**Comment by niboshi** [comment]
Changed the PR title accordingly.

**Comment by niboshi** [comment]
Conflict resolved.

File: tests/chainer_tests/training_tests/extensions_tests/test_snapshot_writers.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+                assert consumer.start.call_count == 1
+                assert q.put.call_count == 3
+                assert q.join.call_count, 1
+                assert consumer.join.call_count == 1
+
+    def test_consume(self):
+        names = [snapshot_writers_path + '.QueueWriter.create_queue',
+                 snapshot_writers_path + '.QueueWriter.create_consumer']
+        with mock.patch(names[0]):
```
Comment (okuta) [inline tests/chainer_tests/training_tests/extensions_tests/test_snapshot_writers.py]: Please more simplified.

File: tests/chainer_tests/training_tests/extensions_tests/test_snapshot_writers.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        self.filename = 'myfile.dat'
+        self.outdir = 'mydir'
+        self.target = mock.MagicMock()
+
+    def test_call(self):
+        q = mock.MagicMock()
+        consumer = mock.MagicMock()
+        names = [snapshot_writers_path + '.QueueWriter.create_queue',
+                 snapshot_writers_path + '.QueueWriter.create_consumer']
+        with mock.patch(names[0], return_value=q):
```
Comment (okuta) [inline tests/chainer_tests/training_tests/extensions_tests/test_snapshot_writers.py]: Please more simplified.

File: chainer/training/extensions/snapshot_writers.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+            filename (str): Name of the file into which the serialized target
+                is saved. It is a concrete file name, i.e. not a pre-formatted
+                template string.
+            outdir (str): Output directory. Corresponds to
+                :py:attr:`Trainer.out <chainer.training.Trainer.out>`.
+            target (dict): Serialized object which will be saved.
+        """
+        raise NotImplementedError
+
+    def __del__(self):
```
Comment (okuta) [inline chainer/training/extensions/snapshot_writers.py]: Is it necessary?

File: tests/chainer_tests/training_tests/extensions_tests/test_snapshot_writers.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+from chainer.training.extensions import snapshot_writers
+
+
+snapshot_writers_path = 'chainer.training.extensions.snapshot_writers'
+
+
+class TestSimpleWriter(unittest.TestCase):
+
+    def setUp(self):
+        self.filename = 'myfile.dat'
```
Comment (okuta) [inline tests/chainer_tests/training_tests/extensions_tests/test_snapshot_writers.py]: Why don't you use temporary directory?

**Comment by niboshi** [comment]
@okuta 
We discussed that the simpler writer interface (rather than separate classes) would be preferred.
Do you have any specific suggestion how the user can specify the writing backend?

**Comment by okuta** [comment]
OK, I understood and respect your discussion.

But, I have one question. Is Process type writer really necessary?
I am concerned that it takes time to transfer data between processes.
I agree to add if it is a function actually used.

**Comment by okuta** [comment]
@niboshi Do you have any comment?

**Comment by niboshi** [comment]
Resolved conflict (merged `snapshot_on_error`), and fixed temporary directory handling in the tests.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by niboshi** [comment]
Ready for merge?

**Comment by okuta** [comment]
OK.
LGTM!

**Pull request merged**